src/stream/src/executor/approx_percentile/global.rs

```diff
@@ -0,0 +1,244 @@
+                        &bucket_state_table,
+                        row_count as u64,
+                        self.quantile,
+                        self.base,
+                    )
+                    .await?;
+                    let percentile_chunk = if row_count_state.is_none() {
+                        StreamChunk::from_rows(
+                            &[(Op::Insert, &[new_output.clone()])],
+                            &[DataType::Float64],
+                        )
+                    } else {
+                        StreamChunk::from_rows(
+                            &[
+                                (Op::UpdateDelete, &[prev_output.clone()]),
+                                (Op::UpdateInsert, &[new_output.clone()]),
+                            ],
+                            &[DataType::Float64],
+                        )
+                    };
+                    prev_output = new_output;
+                    yield Message::Chunk(percentile_chunk);
+
+                    let new_row_count_state = &[Datum::from(ScalarImpl::Int64(row_count))];
+                    if let Some(row_count_state) = row_count_state {
+                        count_state_table.update(row_count_state, new_row_count_state);
+                    } else {
+                        count_state_table.insert(new_row_count_state);
+                    }
+                    row_count_state = Some(new_row_count_state.into_owned_row());
+                    count_state_table.commit(barrier.epoch).await?;
+
+                    bucket_state_table.commit(barrier.epoch).await?;
+
+                    yield Message::Barrier(barrier);
+                }
+                m => yield m,
+            }
+        }
+    }
+
+    /// We have these scenarios to consider, based on row count state.
+    /// 1. We have no row count state, this means it's the bootstrap init for this executor.
+    ///    Output NULL as an INSERT. Persist row count state=0.
+    /// 2. We have row count state.
+    ///    Output UPDATE (`old_state`, `new_state`) to downstream.
+    async fn get_output(
+        bucket_state_table: &StateTable<S>,
+        row_count: u64,
+        quantile: f64,
+        base: f64,
+    ) -> StreamExecutorResult<Datum> {
+        let quantile_count = (row_count as f64 * quantile).floor() as u64;
+        let mut acc_count = 0;
+        let neg_bounds: (Bound<OwnedRow>, Bound<OwnedRow>) = (
+            Bound::Unbounded,
+            Bound::Excluded([Datum::from(ScalarImpl::Int16(0))].to_owned_row()),
+        );
+        let non_neg_bounds: (Bound<OwnedRow>, Bound<OwnedRow>) = (
+            Bound::Included([Datum::from(ScalarImpl::Int16(0))].to_owned_row()),
+            Bound::Unbounded,
+        );
+        // Just iterate over the singleton vnode.
+        // TODO(kwannoel): Should we just use separate state tables for
+        // positive and negative counts?
+        // Reverse iterator is not as efficient.
```

Comment:
TBH I'm not totally sure about this design decision.
The alternative is to split into 3 tables:
- zeros
- neg
- pos

Then we don't have to use reverse iteration.

But for now this design seems more efficient.

Reply:
Can we just use an empty prefix to scan the state table? The order should start from neg to pos.

Reply:
No it doesn't work. Because we aren't just iterating on negative values.

Consider the following is in the state table (with the `count` column elided)

```
| sign | bucket_id |
| - | - |
| -1     | -3 |
| -1      | -1      |
| -1 | 1 |
| -1 | 2| 
| 0 | 0 |
| 1 | 1 |
| 1 | 2 |
```

Our iteration order will be:
```
(-1, -3), (-1,-1), (-1, 1), (-1, 2), (0,0), (+1, 1), (+1, 2)
```

But we actually want
```
(-1, 2), (-1,1), (-1, -1), (-1, -3), (0,0), (+1, 1), (+1, 2)
```

Because

```
- base^2 < - base ^(-3)
```

Reply:
I see. Actually, we can do a math trick here, for example, encoding `bucket_id` as  `- bucket_id` for sign with `-1`. Then we will get the order we want.

Reply:
> I see. Actually, we can do a math trick here, for example, encoding bucket_id as - bucket_id for sign with -1. Then we will get the order we want.

That's a good idea, but I intend to keep this suggestion as a future optimization as well. Currently this is just a few lines in one location:
```
        for keyed_row in bucket_state_table
            .rev_iter_with_prefix(&[Datum::None; 0], &neg_bounds, PrefetchOptions::default())
            .await?
```

Adding `-1` sign, means at all areas where we encode / decode, we will need to add the logic to reverse it. And probably have to add comments there to explain why as well. Here we can just add some comments in a single spot.

Once we add caching, I think we won't do as much rev_iter anymore, and so it will mitigate any performance hit from negative values.